### ISSUE TITLE
Large bundle size with webpack

### ISSUE THREAD

**Issue by nickmccurdy** [puppeteer/replay#257]
## Actual Behavior

I'm making a Chrome recorder extension, and I've noticed the production library bundle was 437KB in my project, which I would consider to be large for a bundle that doesn't need to ship GUI assets.

## Steps to Reproduce the Problem

1. Clone https://gist.github.com/ad7d2641f9b7b8c6012d031edab790c4
1. `npm install`
1. Note the sizes outputted by webpack or on disk

## Specifications

- Version: 0.6.1
- Platform: macOS

**Comment by OrKoN** [comment]
Hey @nickmccurdy I suspect that you don't require Puppeteer for the extension as we currently only support export extensions? In that case, you can safely exclude Puppeteer (which is an optional peer dependency needed for the replay part) from the bundle. Please see this rollup config https://github.com/cypress-io/cypress-chrome-recorder-extension/blob/main/rollup.config.js#L10 as an example in rollup. I believe a similar configuration should be possible for Webpack. Would that work for your extension?

**Comment by nickmccurdy** [comment]
Thanks for the helpful tip, I didn't think about that because I was trying to use tree shaking. Do you think it would make sense to publish with `"sideEffects": false` (which assumes all exported code is pure)?

For anyone else having this issue with webpack, here's the workaround I'm using now:
```js
plugins: [
  new webpack.IgnorePlugin({
    resourceRegExp: /^puppeteer$/,
  })
]
```

**Comment by OrKoN** [comment]
@nickmccurdy unfortunately, I am not familiar with `"sideEffects": false` is it webpack-specific? Does it mean that webpack will be able to strip Puppeteer automatically if the runner extensions are not used by our code?

**Comment by nickmccurdy** [comment]
> unfortunately, I am not familiar with "sideEffects": false. Is it webpack-specific?

Yes. It tells webpack that an entire package's code is pure, so that modules can be tree shaked without losing side effects. However, Rollup can tree shake ES modules.

> Does it mean that webpack will be able to strip Puppeteer automatically if the runner extensions are not used by our code?

In theory, yes, though that should only be used if the entire package is side effect free. If that's not the case, you can use `/*#__PURE__*/` comments on pure code to inform webpack and Rollup that tree shaking that code is safe.

**Comment by OrKoN** [comment]
I gave it a try but it looks it does not work: if I import anything from the replay lib, Webpack decides to import Puppeteer too although Puppeteer is only a dynamic import in the generated ESM code?

**Comment by nickmccurdy** [comment]
Do you have an example webpack config?

**Comment by OrKoN** [comment]
@nickmccurdy I used the one you posted in the gist: https://gist.github.com/nickmccurdy/ad7d2641f9b7b8c6012d031edab790c4

**Comment by nickmccurdy** [comment]
Adding `"sideEffects": false` to `node_modules/@puppeteer/replay/package.json` seems to work, as it results in the `main.js` bundle being empty (which makes sense because we told webpack it has no side effects and we aren't using any imports from it). However, this unfortunately still requires installing `puppeteer` and getting webpack to build it with the fairly large `resolve.fallback` config. I'm hoping someone else has an idea for reducing config, as I'm not sure how to fix it myself.

Working on updating the Gist.
